File: web/src/algorithms/CrossSettlementDfsEngine.ts
import { CrossSettlementOffer, CrossSettlementParticipant, CrossSettlementStep, User } from '../types';

/**
 * Scans for directed debt cycles using DFS algorithm (e.g. A -> B -> C -> A).
 * 1:1 Direct Port from Android Kotlin CrossSettlementDfsEngine.kt
 */
export const CrossSettlementDfsEngine = {
  detectCrossSettlementCycles(
    usersMap: Map<string, User>,
    pairwiseMatrix: Map<string, number> // key: `${from}_${to}` -> amount
  ): CrossSettlementOffer[] {
    const userIds = Array.from(usersMap.keys());
    const graph = new Map<string, Map<string, number>>();

    // Build directed adjacency list: graph[u][v] = amount (u owes v)
    pairwiseMatrix.forEach((amount, key) => {
      if (amount > 0.01) {
        const [from, to] = key.split('_');
        if (!graph.has(from)) {
          graph.set(from, new Map());
        }
        graph.get(from)!.set(to, amount);
      }
    });

    const visited = new Set<string>();
    const recursionStack: string[] = [];
    const detectedCycles: string[][] = [];

    function dfs(current: string) {
      visited.add(current);
      recursionStack.push(current);

      const neighbors = graph.get(current) || new Map<string, number>();
      for (const [nextUser] of neighbors.entries()) {
        if (recursionStack.includes(nextUser)) {
          // Cycle detected!
          const cycleStartIndex = recursionStack.indexOf(nextUser);
          const cycle = recursionStack.slice(cycleStartIndex);
          if (cycle.length >= 3 && !CrossSettlementDfsEngine.isDuplicateCycle(detectedCycles, cycle)) {
            detectedCycles.push(cycle);
          }
        } else if (!visited.has(nextUser)) {
          dfs(nextUser);
        }
      }

      recursionStack.pop();
    }

    for (const u of userIds) {
      if (!visited.has(u)) {
        dfs(u);
      }
    }

    // Convert raw cycle node lists into CrossSettlementOffer domain models
    const offers: CrossSettlementOffer[] = [];

    for (const cycle of detectedCycles) {
      const steps: CrossSettlementStep[] = [];
      let minCapacity = Number.MAX_VALUE;

      for (let i = 0; i < cycle.length; i++) {
        const fromId = cycle[i];
        const toId = cycle[(i + 1) % cycle.length];
        const owedAmount = graph.get(fromId)?.get(toId) || 0;

        if (owedAmount < minCapacity) {
          minCapacity = owedAmount;
        }

        const fromUser = usersMap.get(fromId);
        const toUser = usersMap.get(toId);

        steps.push({
          fromUserId: fromId,
          fromUserName: fromUser?.fullName || fromUser?.username || fromId,
          toUserId: toId,
          toUserName: toUser?.fullName || toUser?.username || toId,
          amount: 0 // Set after bottleneck minCapacity is finalized
        });
      }

      if (minCapacity > 0.01) {
        const finalizedSteps = steps.map((s) => ({ ...s, amount: minCapacity }));
        const participants: CrossSettlementParticipant[] = cycle
          .map((uid) => usersMap.get(uid))
          .filter((u): u is User => !!u)
          .map((u) => ({
            id: u.id,
            name: u.fullName,
            avatar: u.avatarUrl || '👤',
            username: u.username
          }));

        const initialApprovals: Record<string, boolean> = {};
        cycle.forEach((uid) => {
          initialApprovals[uid] = false;
        });

        offers.push({
          id: `offer_${Date.now()}_${Math.random().toString(36).substring(2, 9)}`,
          cycleAmount: minCapacity,
          participants,
          steps: finalizedSteps,
          approvals: initialApprovals,
          status: 'PENDING',
          createdAt: Date.now().toString()
        });
      }
    }

    return offers;
  },

  isDuplicateCycle(existingCycles: string[][], newCycle: string[]): boolean {
    const newSet = new Set(newCycle);
    return existingCycles.some((existing) => {
      if (existing.length !== newCycle.length) return false;
      const existingSet = new Set(existing);
      return Array.from(newSet).every((item) => existingSet.has(item));
    });
  }
};

File: app/src/main/java/com/ardabank/aradapay/domain/algorithm/CrossSettlementDfsEngine.kt
package com.ardabank.aradapay.domain.algorithm

import com.ardabank.aradapay.domain.model.ApprovalStatus
import com.ardabank.aradapay.domain.model.CrossSettlementOffer
import com.ardabank.aradapay.domain.model.CrossSettlementParticipant
import com.ardabank.aradapay.domain.model.CrossSettlementStep
import com.ardabank.aradapay.domain.model.User
import java.util.UUID

object CrossSettlementDfsEngine {

    /**
     * Scans for directed debt cycles using DFS algorithm (e.g. A -> B -> C -> A).
     * Returns a list of CrossSettlementOffer detected.
     */
    fun detectCrossSettlementCycles(
        usersMap: Map<String, User>,
        pairwiseMatrix: Map<Pair<String, String>, Double>
    ): List<CrossSettlementOffer> {
        val userIds = usersMap.keys.toList()
        val graph = mutableMapOf<String, MutableMap<String, Double>>()

        // Build directed adjacency list: graph[u][v] = amount (u owes v)
        pairwiseMatrix.forEach { (pair, amount) ->
            if (amount > 0.01) {
                val (from, to) = pair
                graph.getOrPut(from) { mutableMapOf() }[to] = amount
            }
        }

        val visited = mutableSetOf<String>()
        val recursionStack = mutableListOf<String>()
        val detectedCycles = mutableListOf<List<String>>()

        fun dfs(current: String) {
            visited.add(current)
            recursionStack.add(current)

            val neighbors = graph[current] ?: emptyMap()
            for ((nextUser, _) in neighbors) {
                if (recursionStack.contains(nextUser)) {
                    // Cycle detected!
                    val cycleStartIndex = recursionStack.indexOf(nextUser)
                    val cycle = recursionStack.subList(cycleStartIndex, recursionStack.size).toList()
                    if (cycle.size >= 3 && !isDuplicateCycle(detectedCycles, cycle)) {
                        detectedCycles.add(cycle)
                    }
                } else if (!visited.contains(nextUser)) {
                    dfs(nextUser)
                }
            }

            recursionStack.removeAt(recursionStack.size - 1)
        }

        for (u in userIds) {
            if (!visited.contains(u)) {
                dfs(u)
            }
        }

        // Convert raw cycle node lists into CrossSettlementOffer domain models
        val offers = mutableListOf<CrossSettlementOffer>()

        for (cycle in detectedCycles) {
            val steps = mutableListOf<CrossSettlementStep>()
            var minCapacity = Double.MAX_VALUE

            for (i in cycle.indices) {
                val fromId = cycle[i]
                val toId = cycle[(i + 1) % cycle.size]
                val owedAmount = graph[fromId]?.get(toId) ?: 0.0

                if (owedAmount < minCapacity) {
                    minCapacity = owedAmount
                }

                val fromUser = usersMap[fromId]
                val toUser = usersMap[toId]

                steps.add(
                    CrossSettlementStep(
                        fromUserId = fromId,
                        fromUserName = fromUser?.fullName ?: (fromUser?.username ?: fromId),
                        toUserId = toId,
                        toUserName = toUser?.fullName ?: (toUser?.username ?: toId),
                        amount = 0.0 // Set after bottleneck minCapacity is finalized
                    )
                )
            }

            if (minCapacity > 0.01) {
                val finalizedSteps = steps.map { it.copy(amount = minCapacity) }
                val participants = cycle.mapNotNull { uid ->
                    usersMap[uid]?.let { u ->
                        CrossSettlementParticipant(
                            id = u.id,
                            name = u.fullName,
                            avatar = u.avatarUrl,
                            username = u.username
                        )
                    }
                }

                val initialApprovals = cycle.associateWith { false }

                offers.add(
                    CrossSettlementOffer(
                        id = UUID.randomUUID().toString(),
                        cycleAmount = minCapacity,
                        participants = participants,
                        steps = finalizedSteps,
                        approvals = initialApprovals,
                        status = ApprovalStatus.PENDING,
                        createdAt = System.currentTimeMillis().toString()
                    )
                )
            }
        }

        return offers
    }

    private fun isDuplicateCycle(
        existingCycles: List<List<String>>,
        newCycle: List<String>
    ): Boolean {
        val newSet = newCycle.toSet()
        return existingCycles.any { existing ->
            existing.toSet() == newSet && existing.size == newCycle.size
        }
    }
}
